gradle cleanup

supersaiyansubtlety opened the pull request:
## Ready!

### Tasks
- [x] the thisening
- [x] the finalening
- [x] the line wrappening
- [x] the abstractening
- [x] the registerening
- [x] move all custom tasks to plugins
- [x] move (most) task configuration to plugins (makes it easier to see dependencies and makes tasks more reusable)
- [x] eliminate `buildSrc` referencing `build.gradle` things
- [x] eliminate project access in task actions
- [x] refine inputs/outputs and dependencies
- [x] move more things to `QuiltMappingsExtension`
- [x] separate `QuiltMappingsPlugin`'s contents into several plugins
- [x] wait for #659 which fixes tests, make sure tests still pass
- [x] resolve a few remaining questions
- ~~replace `TasksTest` and depend on gradle test kit~~ maybe in a followup PR
- ~~apply a checkstyle to `buildSrc`?~~ maybe in a followup PR

## Summary

renames `MappingsPlugin`/`mappings-logic` -> `QuiltMappingsPlugin`/`quilt-mappings`
renames `MappingssExtension`/`mappings` -> `QuiltMappingsExtension`/`quiltMappings`
splits `QuiltMappingsPlugin` into:
- QuiltMappingsBasePlugin
- MinecraftJarsPlugin
- MapMinecraftJarsPlugin
- MapV2Plugin
- MapIntermediaryPlugin
- ProcessMappingsPlugin
- EnigmaMappingsPlugin
- TargetDiffPlugin

all implement `MappingsProjectPlugin` and are applied by `QuiltMappingsPlugin`

eliminates `FileConstants`:
- some common directories are in `MappingsProjectPlugin`
- files that were the output of one task are now defined where that task's output is set and other tasks take that output as input

adds several extensions that expose objects created by their plugins
extensions are accessed via plugin instances which ensures extension have been created

no use of `tasks.named(...)` to access custom tasks:
- instead they're accessed via the provider returned by `tasks.register(...)`
- tasks used outside of the plugin they're defined in are accessed via `Tasks` objects obtained via extensions, obtained via plugin instances
- this makes sure required plugins are applied and the tasks are registered

ensures all custom tasks have a group
eliminates `MappingsTask` and `DefaultMappingsTask`
(some `MappingsTask` methods are replaced with helper methods in more specific task interfaces and plugins)

moves `Constants.MINECRAFT_VERSION` to `libs.versions.toml` and passes it to `QuiltMappingsExtension`
moves `Constants.MAPPINGS_VERSION` to direct `project.version` assignment in `build.gradle` and passes it to `QuiltMappingsExtension`
moves dictionary file url to `libs.versions.toml` and downloads it with an ivy repository hack
eliminates `downloadDictionaryFile` and passes the dictionary file to `mappingLint` in  `build.gradle`
saves `unpickVersion` in a `QuiltMappingsEntension` field and passes it to the constructors of tasks that require it
    (it's required at configuration time for some tasks so it can't be a property)

extracts a bunch of magic strings to interfaces in `constants` package
moves all task names to their classes and renames all `TASK_NAME` fields `<ACTUAL_NAME>_TASK_NAME`
adds javadocs to all task names that link the plugin that registers them
adds javadocs to each plugin with any `configurEach`es listing what they do (these can be annoying to track down)
adds javadocs to each class that's `configureEach`ed and to all their subclasses, linking the plugin that does the configuring (these can be really annoying to track down)

moves almost all outputs to `build/*`
moves `combineUnpickDefinitions`'s output to `build/mappings/`
eliminates custom `clean` task logic (everything is in `build/` and gets cleaned)

names `AbstractArchiveTask` outputs using its name composition methods
adds `ArtifactFileProducingTask` which mimics `AbstractArchiveTask` name composition (implemented by `CompressTinyTask`)
eliminates explicit classifiers for published artifacts in `build.gralde`

mapped providers are the solution to every problem

moves `unpick.json` to `unpick/` and unpick definitions to `unpick/definitions/`
moves `enimga_profile.json` and `simple_type_field_names.json5` to `enigma/`

eliminates `decompileClasspath` configuration -> passes `downloadMinecraftLibraries.librariesDir.fileTree`
replaces `CheckStuffExists/Matches` tasks with mapped providers

populates `unpickCli` configuration using included build `:unpick-holders`
splits `enigmaRuntime` configuration into `enigmaSwing` and `enigmaServer`, populates them using `:unpick-holders`

replaces `de.undercouch.gradle.tasks.download.DownloadAction` usage with `DownloadUtil` methods (which use `org.apache.commons.io.FileUtils::copyURLToFile`)
eliminates `de.undercouch:gradle-download-task` dependency

adds `generateDiff` task (Windows users may have to put `diff` on their PATH)

### Future
- use `generateDiff` in `gerenate-diff.yml` (pass output path to avoid magic string)
- move `namedSource/` and `namedTargetSource/`once `generateDiff` is used in `gerenate-diff.yml`
- update/replace licenser plugin: the last release is from 2021 and it's incompatible with gradle 9
- `@CachableTask` everything
- [configuration cache](https://docs.gradle.org/current/userguide/configuration_cache.html)

supersaiyansubtlety committed:
make simple_type_field_names an input of insertAutoGeneratedMappings bettererly
move EnigmaProfile to MappingsExtension

supersaiyansubtlety committed:
replace MappingTask::getTaskByType with MappingsTask::getTaskNamed; tasks should not be retreived by type alone, adding a new task with the same type would break things
(most calls to MappingsTask::getTaskNamed will probably be eliminated when configuration is moved to MappingsPlugin)
the thisening
the abstractening
the finalening
the line wrappening

supersaiyansubtlety commented:
~~Early reviewers beware: I will probably be rewriting history so any review progress on github may be lost.~~
Ready ~~-ish~~

supersaiyansubtlety committed:
the registerening
add missing get() to extractServerJar

supersaiyansubtlety committed:
move downloadVersionsManifest and downloadWantedVersionManifest configuration to MappingsPlugin
refine downloadWantedVersionManifest inputs/outputs

supersaiyansubtlety committed:
move downloadMinecraftJars, extractServerJar, and mergeJars configuration to MappingsExtension
refine their properties and dependencies
make ExtractServerJarTask's getServerBootstrapJar() an input instead of an output
make downloadMinecraftJars not finalizedBy extractServerJar as input/output handling is preferable

supersaiyansubtlety committed:
move downloadMinecraftLibraries and openGlConstantUnpickGenerator configuration to MappingsPlugin
rework both tasks' inputs and outputs significantly
downloadMinecraftLibraries is still screwy though: it accesses and modifies project in its task action
add Task suffix to OpenGlConstantUnpickGenerator class
rename PropertyUtil -> ProviderUtil and add projectDirProviderOf

supersaiyansubtlety committed:
AbstractDownloadMappingsTask -> DownloadMappingsTask, reworked its properties, and separated ExtractTinyMappingsTask
added ExtractTinyMappingsTasks for hashed and intermediary
removed DownloadIntermediaryMappingsTask class, it's just a DownloadMappingsTask now
added Constants.INTERMEDIARY_MAPPINGS_NAME, but haven't replaced duplicates yet
made some convensions in MappingsPlugin to use flatMap instead of lambdas
move MappingsExtension's getEnigmaProfileFile convension to MappingsPlugin

supersaiyansubtlety committed:
move conventions for MergeTinyTask, MergeTinyV2Task, and MergeIntermediaryTask to MappingsPlugin
refine their and their superclass' properties

supersaiyansubtlety committed:
update MapJarTask and subclasses
update CheckTargetVersionExistsTask (will convert to BuildServerice later)
add TargetVersionConsumingTask interface for CheckTargetVersionExistsTask consumers
update CheckUnpickVersionsMatchTask (will convert to BuildServerice later)
add UnpickVersionsMatchConsumingTask for CheckUnpickVersionsMatchTask consumers
separate ExtractTargetMappingJarTask from DownloadTargetMappingJarTask and remove output files that just pointed to individual extracted files
clean up the rest of the diff tasks (still WIP)
move unpickTargetJar from build.gradle to buildSrc and give add a type for it: UnpickTargetJarTask
move remapTargetUnpickDefinitions from build.gradle to buildSrc and add a type for it: RemapTargetUnpickDefinitionsTask
move decompileTargetVineflower from build.gradle to buildSrc and add a type for it: DecompileTargetTask
add more TODOs based on @lukebemish's early review

lukebemish commented on buildSrc/src/main/java/quilt/internal/MappingsExtension.java:
```diff
@@ -1,15 +1,54 @@
+    public MappingsExtension(Project project) {
+        this.fileConstants = new FileConstants(project);
+
+        this.enigmaProfile = this.getEnigmaProfileFile()
```
`map`ed providers _are not cached_, and I assume this is used in several placed  -- so instead of having, say, a `public final Provider<EnigmaProfile> enigmaProfile`, have a `Property<EnigmaProfile>` and `set` it to the mapped provider, which _is_ cached

lukebemish commented on buildSrc/src/main/java/quilt/internal/MappingsPlugin.java:
```diff
@@ -24,66 +31,277 @@
-            task.getMappingDirectory().set(mappingLintTask.getMappingDirectory());
-            mappingLintTask.dependsOn(task);
+    public void apply(@NotNull Project project) {
+        final var ext = project.getExtensions().create("mappings", MappingsExtension.class, project);
```
Not _really_ a big deal, but the `Project` can just be `@Inject`ed here.

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/EnigmaProfileConsumingTask.java:
```diff
@@ -0,0 +1,49 @@
+            "Up-to-date-ness is ensured by getSimpleTypeFieldNamesFiles and its source, " +
+            "MappingsExtension::getEnigmaProfileFile."
+    )
+    public abstract Property<EnigmaProfile> getEnigmaProfile();
```
This is enough for build cache, but won't get you to config cache -- _all_ task properties need to be serialize-able or transient (in which case they may not exist at task runtime), `@Internal` or `@Input`s or `@Output`s or whatever. Question for your consideration: is there ever more than a single `EnigmaProfile` created by this plugin, at all? If the answer is no, use a build service!

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/EnigmaProfileConsumingTask.java:
```diff
@@ -0,0 +1,49 @@
+    @Internal(
+        "An EnigmaProfile cannot be fingerprinted. " +
+            "Up-to-date-ness is ensured by getSimpleTypeFieldNamesFiles and its source, " +
+            "MappingsExtension::getEnigmaProfileFile."
```
Is `MappingsExtension::getEnigmaProfileFile` ever actually marked/added as a task input? Or is the only stuff from it that can affect stuff the simpleTypeFieldNamesFiles?

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/decompile/DecompileTask.java:
```diff
@@ -23,89 +24,69 @@
     }
 
     @Internal
     public AbstractDecompiler getAbstractDecompiler() {
```
Similar to the EnigmaProfile stuff above -- this is not conf cache compatible, and looks like a perfect place to use a build service.

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/setup/DownloadMinecraftJarsTask.java:
```diff
@@ -7,85 +7,93 @@
+    @InputFile
+    public abstract RegularFileProperty getVersionFile();
+
+    @Internal("Fingerprinting is handled by getVersionFile")
```
Once again, remember that conf cache is a thing -- if it's used at task run time, use something serializable instead; if it's not, then making it a transient field _should_ work.

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/setup/DownloadMinecraftLibrariesTask.java:
```diff
@@ -3,88 +3,143 @@
-    private final RegularFileProperty versionFile;
+    public abstract RegularFileProperty getVersionFile();
+
+    @Internal("Fingerprinting is handled by getVersionFile")
```
Same as above

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/setup/DownloadMinecraftLibrariesTask.java:
```diff
@@ -3,88 +3,143 @@
+                // TODO this is screwy.
+                //  Could we put these in a configuration in MappingsPlugin
+                //  and pass the configuration to an input of this task?
+                this.getProject().getDependencies().add("decompileClasspath", library.getName());
```
So, basically -- you cannot, and should not, add dependencies at task execution time! You can expect all dependencies to be set up before then and need to set them up before then. You also should not pass around `Configuration`s as task inputs/properties used at task run time (config cache incompatible if nothing else) If you need this info at that time to do so -- perhaps a `ValueSource` could do it instead of a task? Or, if having it be a task is still easiest, have them saved in a file that's generated by a task ran when _`buildSrc`_ is built, so that they're available in the main project?

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/setup/DownloadWantedVersionManifestTask.java:
```diff
@@ -2,59 +2,84 @@
+    public abstract RegularFileProperty getManifest();
+
+    @Internal("Fingerprinting is handled by getManifest and getReleaseTime")
+    protected abstract Property<VersionEntry> getManifestVersion();
```
Same worry as above.

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/setup/ExtractServerJarTask.java:
```diff
@@ -1,35 +1,37 @@
+        FileUtils.copyFile(
+            // TODO eliminate project access in task action
+            this.getProject()
+                .zipTree(this.getServerBootstrapJar().get())
```
Ah yes, our good friend `zipTree` that for some reason is not available on any non-internal injectable service -- this sucks unfortunately, and there's no nice solution other than building your own zip utility or splitting the task. Or possibly doing it in the task ctor? No clue if that laziness is lazy enough to work fine.

supersaiyansubtlety replied:
Yeah I've updated another task to save the `zipTree` in the constructor and it's working fine

supersaiyansubtlety replied:
The issue I asked about on discord was because the other task was also *outputting* the file that was getting unzipped, so I had to separate the unzipping to another task.

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/setup/DownloadMinecraftLibrariesTask.java:
```diff
@@ -3,88 +3,143 @@
+    @OutputDirectory
+    public abstract DirectoryProperty getLibrariesDir();
+
+    @Internal("Fingerprinting is handled by getLibrariesDir")
```
Same as above

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/unpick/gen/OpenGlConstantUnpickGeneratorTask.java:
```diff
@@ -0,0 +1,389 @@
+    public abstract RegularFileProperty getPerVersionMappingsJar();
+
+    @Input
+    public abstract MapProperty<String, File> getArtifactsByUrl();
```
I have a sneaking suspicion this is meant to be `@InputFiles`, if you're using the artifacts' contents.

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/diff/CheckTargetVersionExistsTask.java:
```diff
@@ -9,50 +9,54 @@
+public abstract class CheckTargetVersionExistsTask extends DefaultMappingsTask {
     public static final String TASK_NAME = "checkTargetVersionExists";
 
     @Internal
```
Uh... I don't like whatever this is doing. Does it set the field _during task execution_ then expect tasks running after it to grab the field from the task and use it? Iffy, to say the least, though I suppose functional. Regardless -- I'd annotate the method, not the field, here in either case.

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/diff/CheckUnpickVersionsMatchTask.java:
```diff
@@ -12,11 +12,11 @@
+    public abstract RegularFileProperty getUnpickJson();
 
     @Internal
     private boolean match = false;
```
Same as my worry with targetVersion above -- I'm not quite sold on this approach, it seems like a _bit_ of a bad pattern.

lukebemish commented on buildSrc/src/main/java/quilt/internal/tasks/setup/CheckIntermediaryMappingsTask.java:
```diff
@@ -9,7 +9,10 @@
+public abstract class CheckIntermediaryMappingsTask extends DefaultMappingsTask {
     public static final String TASK_NAME = "checkIntermediaryMappings";
 
     @Internal
```
Ooh, here's another of whatever-the-heck these are!

lukebemish reviewed (commented):
Looks pretty good! Unsure if the eventual aim is config cache here (and maybe it ought to be, as config cache will be on by default in gradle 9 which is on the horizon) -- regardless, if so there's a bit of work to do, since a lot of stuff is still kinda unidiomatic in that regard; all things told though a _drastic_ improvement. Didn't comment on any remaining TODOs.

supersaiyansubtlety commented:
~~Note: Rather than marking this PR as ready for review, when it's ready I'll be closing it and creating a new one.~~
~~That will clean up the messy history here on GitHub caused by rewriting history.~~
Never mind, it cleaned itself up, I underestimated GitHub.

supersaiyansubtlety committed:
add missing .get().getAsFile()s to fix genFakeSource and decompileVineflower
temporarily make remapTargetMinecraftJar directly depend on unpickTargetJar, shouldn't be necessary once CheckTargetVersionExists and CheckUnpickVersionsMatchTask are converted to BuildService's
add more TODOs to build.gradle

supersaiyansubtlety committed:
move combineUnpickDefinitions and remapUnpickDefinitions configuration from constructor/build.gradle to MappingsPlugin
move constantsJar from build.gradle to MappingsPlugin
move unpickHashedJar from build.gradle to MappingsPlugin
move v2UnmergedMappingsJar, v2MergedMappingsJar, intermediaryV2MappingsJar, and intermediaryV2MergedMappingsJar to MappingsPlugin
reorder some tasks in MappingsPlugin to accomodate new tasks

supersaiyansubtlety committed:
move generated unpick definition outputs to fileConstants.buildDir
move unpick.json to unpick/ and unpick definitions to unpick/definitions/
add UnpickGenTask::getGeneratedUnpickDefinitions
refine CombineUnpickDefinitionsTask and its configuration
add unpick/definitions/ to CombineUnpickDefinitionsTask's input in build.gradle
rename OpenGlConstantUnpickGeneratorTask -> OpenGlConstantUnpickGenTask
make unpick.json's location configurable via MappingsExtension and configure it in build.gradle
remove all unpick-related files from FileConstants (obsolete, or only used once in which case inlined)

supersaiyansubtlety committed:
rename MappingsPlugin -> QuiltMappingsPlugin and its id mappings-logic -> quilt-mappings
rename MappingsExtension -> QuiltMappingsExtension and its name mappings -> quiltMappings
move enimga_profile.json and simple_type_field_names.json5 to enigma/
rename some things in MappingsExtension
refine MappingsV2JarTask's properties, extract some magic constants, and move some configuration to a configureEach in QuiltMappingsPlugin

supersaiyansubtlety committed:
move mappings, mappingsUnpicked, mappingsServer, and mappingsUnpickedServer to QuiltMappingsPlugin (and fix for new enigma profile location)
extract AbstractEnigmaMappingsTask as a superclass of EnigmaMappingsTask and EnigmaMappingsServerTask
allow specifying EnigmaMappingsServerTask optional properties on task invocation
move AbstractEnigmaMappingsTasks' configuration to QuiltMappingsPlugin (except getMainClass configuration)
expose QuiltMappingsExtension::getEnigmaProfileConfig so it can be passed to AbstractEnigmaMappingsTasks
add enigma-server dependency to buildSrc (as part of bundles.enigma.full)
convert EnigmaProfileConsumingTask to an interface
clean up MappingsTask a little

supersaiyansubtlety committed:
improve unpickVersion handling
eliminate VersionCatalog methods from MappingsTask and add TODOs

supersaiyansubtlety committed:
move the following tasks' configuration to QuiltMappingsPlugin and refine their properties:
- buildMappingsTiny
- invertPerVersionMappings
- removeIntermediary
- dropInvalidMappings
- generatePackageInfoMappings

move downloadDictionaryFile configuration to build.gradle (because it's prone to change) and refine its input
move decompileVineflower to QuiltMappingsPlugin
move eraseByteCode and genFakeSource to QuiltMappingsPlugin and give them custom classes

add MappingsDirConsumingTask
add MappingsDirOutputtingTask and implement it on GeneratePackageInfoMappingsTask

eliminate all FileConstants except tempDir which needs more investigation
eliminate RegularFile implementers in QuiltMappingsPlugin
rework JavadocProvider consumption
simplify UnpickGenTask and javadoc it saying they should only output unpick files
remove _GROUP suffix from Constants.Groups fields

supersaiyansubtlety committed:
move tinyJar configuration to QuiltMappingsPlugin
move compressTiny configuration to QuiltMappingsPlugin and use try-with-resources in task action

supersaiyansubtlety committed:
remove CheckIntermediaryMappingsTask and DownloadIntermediaryMappingsTask
extractTinyIntermediaryMappings gets input directly from the intermediary configuration and dependant tasks check if their inputs exist
extract IntermediaryMappingsV2JarTask and check if input exists

supersaiyansubtlety committed:
remove downloadPerVersionMappings and DownloadMappingsTask
extractTinyPerVersionMappings gets input directly from the hashed configuration

supersaiyansubtlety committed:
refine TransformJarClassesTask properties
misc minor cleanup

supersaiyansubtlety committed:
move tempDir to build/ and its definition to QuiltMappingsPlugin
rename cacheFilesMinecraft minecraftDir and move it to build/
eliminate FileConstants
eliminate custom clean task logic
revise insertAutoGeneratedMappings/AddProposedMappingsTask outputs and file names
revise TargetVersionConsumingTask convension providers

supersaiyansubtlety committed:
rework CheckTargetVersionExistsTask -> DownloadTargetMetaFileTask
the target version is obtained via provideTargetVersion

supersaiyansubtlety committed:
rework CheckUnpickVersionsMatchTask -> UnpickVersionsMatchSource

supersaiyansubtlety committed:
make MapJarTask::getAdditionalMappings an input
cleanup remapTargetMinecraftJar
cleanup completed TODOs and commented code in QuiltMappingsPlugin

supersaiyansubtlety committed:
remove @Optional from TargetVersionConsumingTask's inputs since only getTargetVesion's presence is checked
other misc cleanup

supersaiyansubtlety committed:
replace intermediary task input exists checks with IntermediaryDependantTask interface
they only run if intermediary resolves
this ensures intermediary tasks don't run if intermediary doesn't exist but old input files exist
add ExtractTinyIntermediaryMappingsTask for extractTinyIntermediaryMappings
fix AddProposedMappingsTask accidentally creating directories with output file names instead of their parent directories

supersaiyansubtlety committed:
replace AbstractDecompiler's Project with a Logger
the logger was all the project was used for and the decompiler shouldn't access the project
rename DecompilerProvider -> Decompilers.Factory and encapsulate it more

supersaiyansubtlety committed:
extract MappingsTask::startDownload into DownloadTask
replace usage of MappingsTask::outputsNeverUpToDate with @DisableCachingByDefault

supersaiyansubtlety committed:
update CombineUnpickDefinitionsTask's and UnpickGenTask's javadocs
make UnpickGenTask extend MappingsTask

supersaiyansubtlety committed:
rename DownloadImmediate.Builder -> Downloader and eliminate DownloadImmediate
make Downloader#src a URL and eliminate use of deprecated URL constructor
add TODO to consider eliminating dependence on DownloadAction because of project access

supersaiyansubtlety committed:
extract 'diff' group to Constants.Groups.DIFF

supersaiyansubtlety committed:
extract common logic of ExtractServerJarTask, ExtractTinyMappingsTask, and ExtractTargetMappingJarTask
into AbstractExtractZipTask, ExtractSingleZippedFileTask, and ExtractZippedFilesTask

supersaiyansubtlety committed:
add custom task classes for QuiltMappingsPlugin tasks that lack them
move all task names to their classes and rename all 'TASK_NAME' fields '<ACTUAL_NAME>_TASK_NAME'
ensure all QuiltMappingsPlugin tasks implement MappingsTask and have a group
move unpick configuration creation to QuiltMappingsPlugin, extract Constants.UNPICK_NAME
use unpick Main class name instead of string in UnpickJarTask
implement MappingsDirConsumingTask on FindDuplicateMappingFilesTask and use task's logger instead of static logger instance

supersaiyansubtlety committed:
add DownloadUtils which uses FileUtils::copyURLToFile
rewrite download tasks to use DownloadUtils instead of DownloadAction
extract SimpleDownloadTask from tasks that download a single file from a url
move DownloadDictionaryFileTask's configuration to build.gradle (depends on remote that's prone to change)
remove de.undercouch:gradle-download-task dependency
eliminate decompileClasspath configuration and its screwiness in DownloadMinecraftLibrariesTask
cache version info parsed from wanted version manifest in VersionDownloadInfo property and pass that to VersionDownloadInfoConsumingTasks
add SerializableVersionEntry so it can be provided to downloadWantedVersionManifest and carry dependency info
move constants source set definition back to build.gradle
move constantsJar source set input configuration back to build.gradle
move sourcesJar task registration to QuiltMappingsPlugin and give it its own class but keep its source set configuration in build.gradle
move javadocJar task registration to QuiltMappingsPlugin and give it its own class
move some of javadoc's configuration to QuiltMappingsPlugin
extract Constants.Groups.JAVADOC_GENERATION
@Inject an ObjectFactory into MappingsTasks
fix DecompileTask's output being a RegularFileProperty instead of a DirectoryProperty

supersaiyansubtlety committed:
@Inject and use ArchiveOperations in AbstractExtractZipTask

supersaiyansubtlety committed:
move licenser plugin info to libs.versions.toml
eliminate project access in javadoc task doLast
extract IntermediaryMappingsV2JarTask classifiers
move check and jar task configuration from build.gradle to QuiltMappingsPlugin

supersaiyansubtlety committed:
make simple_type_field_names an input of insertAutoGeneratedMappings bestly

supersaiyansubtlety committed:
extract "net/minecraft/unused/packageinfo/" constant
bring the finalening, the thisening, and the line wrappening to the lint package
slightly refactor SpellingChecker::checkPlural

supersaiyansubtlety committed:
replace MappingsJavadocProvider.Source with mapped providers so they carry dependency info

supersaiyansubtlety committed:
extract mappings namespace constants
import Constants.Groups

supersaiyansubtlety committed:
add generateDiff task (not yet used in generate-diff.yml)

supersaiyansubtlety committed:
add quilt/internal/plugin package
extract MappingsProjectPlugin interface
extract MinecraftJarsPlugin and QuiltMappingsBasePlugin

supersaiyansubtlety committed:
extract the following plugins:
- MapMinecraftJarsPlugin
- MapV2Plugin
- MapIntermediaryPlugin
- ProcessMappingsPlugin
- MappingsVerificationPlugin

move unpickVersion to QuiltMappingsExtension

to minimize spaghetti and ensure required plugins are applied:
- access QuiltMappingsExtension via QuiltMappingsBasePlugin's getExt()
- access tasks registered by other plugins via TaskedMappingsProjectPlugin::getTasks

move some file naming methods to new FileUtil class

update and add javadocs

supersaiyansubtlety committed:
extract magic classifier constants

supersaiyansubtlety committed:
improve DiffDirectoriesTask

supersaiyansubtlety committed:
use 'all' gradle distribution

supersaiyansubtlety committed:
add ArtifactFileProducingTask and implement it on CompressTinyTask

supersaiyansubtlety committed:
name AbstractArchiveTask subclass tasks using its archive name composition methods
make the default ArtifactFileTask destination build/libs
eliminate MappingsProjectPlugin#getLibsDir (its already the default for all tasks it was used for)
add QuiltMappingsArtifactTask, implement it on all tasks that used ARCHIVE_FILE_NAME_PREFIX, and configure implementing tasks' baseName and version in QuiltMappingsBasePlugin
improve some javadocs

supersaiyansubtlety committed:
rename package tasks -> task

supersaiyansubtlety committed:
improve javadocs, esp. for configurEach

supersaiyansubtlety committed:
extract EnigmaMappingsPlugin and TargetDiffPlugin
QuiltMappingsPlugin now applies all other MappingsProjectPlugins and does nothing else
replace UnpickVersionsMatchSource with a mapped provider so dependency info is carried
more javadoc adjustments

supersaiyansubtlety committed:
use JavaExec::getArgumentProviders instead of overriding task action in AbstractEnigmaMappingsTask, EnigmaMappingsServerTask, and UnpickJarTask

supersaiyansubtlety committed:
replace VersionDownloadInfo with VersionParser

supersaiyansubtlety committed:
add artifact helper methods to ArtifactFileTask
remove explicit classifiers from artifact configuration in build.gradle (classifiers are extracted from the input)
fix intermediary publishing (broken since the removal of checkIntermediaryMappings)
cleaned up some TODOs

supersaiyansubtlety committed:
replace Constants.MINECRAFT_VERSION with libs version and pass it to QuiltMappingsExtension
replace Constants.MAPPINGS_VERSION with direct assignment to project.version and pass it to QuiltMappingsExtension
add minecraftVersion input to DownloadTargetMetaFileTask
remove provideVersionEntry from DownloadVersionsManifestTask and instead provide it directly in MinecraftJarsPlugin
add minecraftVersion param to SerializableVersionEntry::of
replace MappingsProjectPlugin::provideMappingsDest with provideBuildMappingsDirFile

supersaiyansubtlety committed:
move hashed/hashed-unpicked/named jars to build/minecraft/mapped/
move merged jar to build/minecraft/
add more helper methods to MappingsProjectPlugin

supersaiyansubtlety committed:
move .gradle/targets -> build/targets
extract more magic strings
add more helper methods for common file naming schemes

supersaiyansubtlety committed:
convert Constants to an interface and move it to new constants package
promote all inner Constants classes to the constants package
extract more classifiers and move all classifier constants to constants.Classifiers
extract more extensions and move all extension constants to constants.Extensions

supersaiyansubtlety committed:
rename configuration unpick -> unpickCli
split configuration enigmaRuntime -> enigmaSwign + enigmaServer
add :classpath-holders project and includeBuild it in buildSrc
populate enigmaSwing, enigmaServer and unpickCli with :classpath-holders capabilites so :quilt-mappings build.gradle doesn't have to (thanks to @lukebembish)

supersaiyansubtlety committed:
add convenient hack to ArtifactFileTask

supersaiyansubtlety committed:
provide default dictionary file name in MappingsVerificationPlugin instead of naming it after the revision in build.gradle

supersaiyansubtlety committed:
remove @DisableCachingByDefault from MappingsV2JarTask because it's already on its superclass

supersaiyansubtlety committed:
categorize TODOs

supersaiyansubtlety committed:
DownloadMinecraftLibrariesTask:
- enable caching
- delete old libraries and try downloading new libraries

supersaiyansubtlety committed:
DownloadVersionsManifestTask: replace @DisableCachingByDefault with UntrackedTask so it always re-downloads

supersaiyansubtlety committed:
DownloadWantedVersionManifestTask: remove release time input; obsoleted by other inputs

supersaiyansubtlety committed:
OpenGlConstantUnpickGenTask:
- only take lwjgl file as input instead of all artifacts by name
- remove onlyIf: if the inputs change, the outputs may change

supersaiyansubtlety committed:
remove DownloadMinecraftJarsTask's outsputs upToDateWhen checksum check: it was unecessary and caused outputs to never be up to date

supersaiyansubtlety committed:
recategorize TODO for javadoc

supersaiyansubtlety committed:
replace DownloadVersionsManifestTask with SerializableVersionEntry.Source because output can't be cached

supersaiyansubtlety committed:
replace DownloadDictionaryFileTask with ivy repository hack in build.gradle

supersaiyansubtlety committed:
merge MappingsVerificationPlugin into MapMinecraftJarsPlugin

supersaiyansubtlety committed:
enable caching in gradle.properties

supersaiyansubtlety committed:
remove cache limitation on javadoc task

supersaiyansubtlety commented on buildSrc/src/main/java/quilt/internal/util/DownloadUtil.java:
```diff
@@ -0,0 +1,98 @@
+                logger.lifecycle("Downloading {}", url);
+            }
+
+            // TODO QUESTION I just made these timeouts up, they could probably be better
```
Suggestions?

supersaiyansubtlety replied:
these have worked so far, we can lengthen them later if needed

supersaiyansubtlety commented on buildSrc/src/main/java/quilt/internal/task/decompile/DecompileVineflowerTask.java:
```diff
@@ -0,0 +1,34 @@
+    public void decompile() throws IOException {
+        FileUtils.deleteDirectory(this.getOutput().get().getAsFile());
+
+        // TODO QUESTION this sometimes logs a non-fatal exception:
+        //  java.nio.file.FileSystemAlreadyExistsException
+        //  it's thrown from
+        //  org.jetbrains.java.decompiler.main.plugins.JarPluginLoader:init
+        //  it seems harmless
```
Please verify this is harmless

supersaiyansubtlety replied:
hasn't cause any issues afaict
I've left the comment but remove the TODO

supersaiyansubtlety commented on buildSrc/src/main/java/quilt/internal/task/ArtifactFileTask.java:
```diff
@@ -0,0 +1,163 @@
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.OutputFile;
+
+// TODO QUESTION Is there a way to make it so MavenPublication#artifact will accept
+//  these tasks directly and run them to build their outputs, similar to AbstractArchiveTask?
+//  (eliminating the need for the artifact convenience methods)
+//  I considered implementing PublishArtifact on this, but it's annotated with @HasInternalProtocol.
+//  This suggests that it would work but doesn't recommend it:
+//  https://github.com/gradle/gradle/issues/17273#issuecomment-858400396
```
Suggestions?

supersaiyansubtlety commented on buildSrc/src/main/java/quilt/internal/plugin/MapV2Plugin.java:
```diff
@@ -0,0 +1,261 @@
+            MergeTinyV2Task.MERGE_TINY_V_2_TASK_NAME,
+            MergeTinyV2Task.class,
+            task -> {
+                // TODO QUESTION this used to be dependent on v2UnmergedMappingsJar, but afaict it has no effect on this task
```
Please check this

OroArmor replied:
Yarn doesnt have `v2UnmergedMappingsJar` anymore, its probably fine

supersaiyansubtlety reviewed (commented):
A few remaining questions:

resolved: probably not much benefit, but maybe worth doing in a followup PR
~~Should I just throw `@CachableTask` on everything?~~
~~Do we benefit much from build cache without it?~~

resolved: seems fine
~~Please check the last several commits where I messed with the caching of a few tasks.~~

Also please manually test this into oblivion.

all resolved
~~TODO QUESTIONS:~~

OroArmor commented on build.gradle:
```diff
@@ -1,403 +1,178 @@
 
-	// verbose = true // enable to debug
+	// enable to debug
+	final debug = false
```
Can we make this a property? maybe something like `-Dquilt.mappings.debug.javadoc`?

OroArmor commented on buildSrc/src/main/java/quilt/internal/constants/Namespaces.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface Namespaces {
+    String INTERMEDIARY = Constants.INTERMEDIARY_MAPPINGS_NAME;
+    String PER_VERSION = Constants.PER_VERSION_MAPPINGS_NAME;
```
I would make this the source of truth and have the others pull from this. Also I would move intermediary to the bottom. We might also want a rename of `per_version` to `intermediate` (not intermediary, thanks fabric), since the idea was to make the QM buildscript mostly independent of the middle layer for the mappings

OroArmor commented on buildSrc/src/main/java/quilt/internal/task/EnigmaProfileConsumingTask.java:
```diff
@@ -0,0 +1,42 @@
+     * so they must be considered for up-to-date checks.
+     */
+    @InputFiles
+    ConfigurableFileCollection getSimpleTypeFieldNamesFiles();
```
Could we change this to something like `getProfileFileDependencies` or something? i have a plugin PR that adds another file to this list and this name is somewhat confusing

OroArmor commented on buildSrc/src/main/java/quilt/internal/task/MappingsTask.java:
```diff
@@ -0,0 +1,9 @@
+/**
+ * All tasks added by {@link QuiltMappingsPlugin QuiltMappingsPlugin} share this type.
+ */
+public interface MappingsTask extends Task { }
```
Do we need this anymore or at all?

supersaiyansubtlety replied:
Not really, I kept it in case we wanted to add commonly injected services or broad convenience methods similar to `MappingsProjectPlugin`, or in case we want to identify plugin tasks by type.

But since the PR is basically ready and this isn't used I think that's unlikely.

If we remove it we can also remove `DefaultMappingsTask` and just set groups in each task.

supersaiyansubtlety replied:
eliminated both

OroArmor commented on unpick-definitions/.gitignore:
```diff
@@ -1,4 +0,0 @@
-# Put the auto generated files here
```
Were these moved somewhere else?

supersaiyansubtlety replied:
`build/mappings/`:
https://github.com/supersaiyansubtlety/quilt-mappings/blob/a1c4c484c3aa329859f59ca6961ac40700eafad8/buildSrc/src/main/java/quilt/internal/plugin/MapV2Plugin.java#L130-L136

OroArmor reviewed (changes requested):
Few nitpicks, WAY larger than i expected, but looks *amazing*. great job!

supersaiyansubtlety commented:
resolved
~~Forgot one question:~~

~~These properties should be versions in the catalog, but `gson` and `netty` already have different, unused versions in the catalog:~~
- ~~https://github.com/supersaiyansubtlety/quilt-mappings/blob/a1c4c484c3aa329859f59ca6961ac40700eafad8/gradle.properties#L8-L12~~
- ~~https://github.com/supersaiyansubtlety/quilt-mappings/blob/a1c4c484c3aa329859f59ca6961ac40700eafad8/gradle/libs.versions.toml#L22-L23~~

~~Which versions of `gson` and `netty` do we want?~~

supersaiyansubtlety committed:
add quilt.mappings.debug.javadoc property

supersaiyansubtlety committed:
clean DecompileVineflowerTask's output dir instead of deleting it

supersaiyansubtlety committed:
remove TODO QUESTION about mergeTinyV2's old dependency on v2UnmergedMappingsJar

supersaiyansubtlety committed:
rename getSimpleTypeFieldNamesFiles -> getProfileFileDependencies

supersaiyansubtlety committed:
make Namespaces PER_VERSION and INTERMEDIARY the sources of truth for their strings, eliminate them from Constants

supersaiyansubtlety committed:
rename all constants containing PER_VERSION -> INTERMEDIATE
rename tasks:
- mapPerVersionMappingsJar -> mapIntermediateMappingsJar
- invertPerVersionMappings -> invertIntermediateMappings
- extractTinyPerVersionMappings -> extractTinyIntermediateMappings

supersaiyansubtlety committed:
increase gradle memory from 3Gb to 4Gb because decompileTargetVineflower ran out of memory

supersaiyansubtlety commented:
Re: [Configuration cache](https://docs.gradle.org/current/userguide/configuration_cache.html)

I tried running some tasks with `--configuration-cache` and got:
> Configuration cache state could not be cached: field `pendingMap` of `org.gradle.api.internal.DefaultNamedDomainObjectCollection$UnfilteredIndex` bean found in field `delegate` of `org.gradle.api.internal.DefaultNamedDomainObjectCollection$FilteredIndex` bean found in field `index` of `org.gradle.api.internal.tasks.DefaultTaskCollection` bean found in field `delegate` of `org.gradle.api.internal.tasks.DefaultRealizableTaskCollection` bean found in input property `$1` of task `:buildMappingsTiny` of type `quilt.internal.task.build.BuildMappingsTinyTask`: error writing value of type 'java.util.LinkedHashMap'
> java.util.ConcurrentModificationException (no error message)

That's for `buildMappingsTiny`, I got a similar error mentioning `error writing value of type 'java.util.LinkedHashMap'` for `findDuplicateMappingFiles`.

idk how to fix these because idk where the `LinkedHashMap` comes from.

configuration caching will have to go in a followup PR
~~If it turns out it wouldn't take too many changes to fix config caching I'm ok with including it in this PR, otherwise it'll have to go in a followup PR.~~

supersaiyansubtlety committed:
add libs.enigma.plugin dependency back to enigmaSwing and enigmaServer classpaths

eristhea commented:
is there anything that remains to be done on this now that #659 is merged?

supersaiyansubtlety commented:
> is there anything that remains to be done on this now that #659 is merged?

~~It's not waiting on any other PRs, but I need to merge changes and make sure tests work.~~ done

~~There are also a few unresolved questions above.~~ resolved

supersaiyansubtlety committed:
Merge branch '24w45a' into procrastination+gradle-cleanup

supersaiyansubtlety committed:
Merge branch '24w45a' into procrastination+gradle-cleanup (with 24w45a re merge)

supersaiyansubtlety committed:
bring the thisening, the bracketing, the line wrappening, and the finalening to TasksTest and TestUtil
other minor cleanup

supersaiyansubtlety committed:
eliminate MappingsTask

supersaiyansubtlety committed:
eliminate DefaultMappingsTask

supersaiyansubtlety committed:
utilize streams better in UnpickJarTask

supersaiyansubtlety committed:
replace Default/TaskedMappingsProjectPlugin/DefaultTaskedMappingsProjectPlugin with DefaultExtensionedMappingsProjectPlugin + Default/TaskedExtension
access Tasks objects and MapIntermediaryPlugin's intermediaryFile via extension instances obtained via plugin instances instead of directly via plugin instances

supersaiyansubtlety committed:
make QuiltMappingsExtension local variable names clearer and consistent

OroArmor reviewed (approved):
Honestly too big to properly review now. From my previous looks its going great. I trust that you did it correctly

supersaiyansubtlety committed:
ArtifactFileTask:
- remove #getArtifact method as the classifier got stripped when publishing
- make #artifact method add the classifier

supersaiyansubtlety committed:
obtain gson, commons-codec, and commons-compress versions for javadoc from DownloadWantedVersionManifestTask#provideVersionParser
use netty version from libs.version.toml because minecraft's netty version doesn't have javadocs

supersaiyansubtlety committed:
clean up resolved TODOs

supersaiyansubtlety committed:
re-remove asm snapshot repo which snuck back in during a merge

supersaiyansubtlety committed:
use task logger instead of project logger in Javadoc

supersaiyansubtlety committed:
Merge branch '24w46a' into procrastination+gradle-cleanup

supersaiyansubtlety committed:
put CacheableTask on DownloadWantedVersionManifestTask, DownloadMinecraftJarsTask, and DownloadMinecraftLibrariesTask
closes #492

supersaiyansubtlety committed:
Merge branch '1.21.4-pre1' into procrastination+gradle-cleanup

supersaiyansubtlety committed:
Merge branch '1.21.4-pre1' into procrastination+gradle-cleanup
